### DISCUSSION TITLE
Writing a plugin in Zig

### DISCUSSION THREAD

**Discussion by Khalid-Nowaf** [tenzir/tenzir#3878] [Ideas & Features]
Would it be possible to write a plugin in Zig language?

**Comment by dominiklohmann** [comment]
Hey and welcome to Tenzir!

There's nothing stopping you from doing so on a technical level. The plugin entrypoint is a set of functions with C linkage with fixed names, and for C++ we offer a convenience macro in libtenzir and build scaffolding in our CMake tooling that make it a lot easier to write a plugin.

When you want to create a plugin in Zig, you either have to recreate what our convenience macro and CMake code do in Zig, or write the plugin entrypoint in C++ and link it against a library written in Zig that you call into.

**Reply by Khalid-Nowaf** [reply]
Thanks for the quick answer!

I will try to first option, and it would be a great way to learn more about Tenzir. 

I'm very excited btw, Tenzir looks very promising.

**Reply by dominiklohmann** [reply]
For reference, take a look at the [`TenzirRegisterPlugin` CMake function](https://github.com/tenzir/tenzir/blob/6d6f84e7a3e1816ce58a9ca32faee16b4bc7b17a/cmake/TenzirRegisterPlugin.cmake#L377-L388) and the various definitions of the [`TENZIR_REGISTER_PLUGIN` C++ macros](https://github.com/tenzir/tenzir/blob/6d6f84e7a3e1816ce58a9ca32faee16b4bc7b17a/libtenzir/include/tenzir/plugin.hpp#L1048-L1159).

If you have any questions, feel free to ask. We also have a [community Discord](https://docs.tenzir.com/discord) in which all devs here are quite responsive. I'd be very curious to learn more about your use cases.

**Reply by Khalid-Nowaf** [reply]
Thanks for the reference @dominiklohmann! 

I will keep you posted on Discord.
